add tokenomics to application section (#1729)
hyperledger-labs/splice: docs/src/app_dev/app_val_tokenomics/feat_app_act_marker_nomics.rst

..
   Copyright (c) 2024 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
..
   SPDX-License-Identifier: Apache-2.0

.. _feat_app_act_marker_nomics:

Featured Application Activity Marker
=====================================

Featured application activity markers
(`FeaturedAppActivityMarker <https://docs.dev.sync.global/app_dev/api/splice-api-featured-app-v1/Splice-Api-FeaturedAppRightV1.html#splice-api-featuredapprightv1>`__)

> are created in a transaction for activity on the network that adds value

Reviewer: ```suggestion
can be created to record activity that adds value
```

Author: Done

> but does not involve a CC Transfer (i.e., transfers a RWA like a USDC).

Reviewer: ```suggestion
but does not involve a CC Transfer (e.g., a stable coin transfer or the settlement of a trade).
```

A ``FeaturedAppActivityMarker`` is immediately converted into an

> ``AppRewardCoupon`` by the automation run by the SuperValidators and the

Reviewer: ```suggestion
``AppRewardCoupon`` by the automation run by the Super Validators and the
```

``AppRewardCoupon`` is minted just like in a :ref:`cc_xfer_nomics`.
Please review the application activity marker
`CIP <https://github.com/global-synchronizer-foundation/cips/blob/main/cip-0047/cip-0047.md>`__
for details. A summary follows.

There are two prerequisites for an application to create a
``FeaturedAppActivityMarker``. The first is to become an approved featured
application which was described in the
:ref:`app_val_tokenomics`
section. The second is to update the application code:

   1. Find the fully qualified interface ID for
   `FeaturedAppRight <https://docs.dev.sync.global/app_dev/api/splice-api-featured-app-v1/Splice-Api-FeaturedAppRightV1.html#type-splice-api-featuredapprightv1-featuredappright-34177>`__
   which is
   ``7804375fe5e4c6d5afe067bd314c42fe0b7d005a1300019c73154dd939da4dda:Splice.Api.FeaturedAppRightV1:FeaturedAppRight``
   for ``Splice.Api.FeaturedAppRightV1``. The command ``daml damlc

>    inspect-dar`` can be used to find this.

Reviewer: This is a stable constant. It's probably easiest to just inline this in the follow-up query that you show; and just add a comment that `78043...` is the package-id of the interface definition for the `FeaturedAppRight` interface.

Author: Done but didn't inline.

   2. Query the ledger using this ID to retrieve contracts from the Daml
   ledger that implement the
   `FeaturedAppRight <https://docs.dev.sync.global/app_dev/api/splice-api-featured-app-v1/Splice-Api-FeaturedAppRightV1.html#type-splice-api-featuredapprightv1-featuredappright-34177>`__
   interface. See using
   `toInterface <https://docs.digitalasset.com/build/3.3/reference/daml/interfaces.html#tointerface>`__
   or
   `toInterfaceContractId <https://docs.digitalasset.com/build/3.3/reference/daml/interfaces.html#tointerfacecontractid>`__

>    for more information.

Reviewer: These are Daml functions, which can't be used at the Ledger API level.

What you want to do is to use an interface filter. Here's an example of one from https://github.com/DACH-NY/canton-network-circle/blob/fb91ea6506cff1277328bbfe95dadf219cb425cf/scripts/json-api-v2-example/6-query-holding.sh#L20-L41
```
  curl_check "http://$APP_USER_PARTICIPANT/v2/state/active-contracts" \
    "$token" "application/json" \
      --data-raw '{
      "filter": {
        "filtersByParty": {
            "'$holderParty'": {
              "cumulative": 
              [
                {
                  "identifierFilter": {
                    "InterfaceFilter": {
                      "value": {
                        "interfaceId": "'$holdingInterfaceId'",
                        "includeInterfaceView": true,
                        "includeCreatedEventBlob": false
                      }
                    }
                  }
                }
              ]}        
        }
      },
      "verbose": false,
      "activeAtOffset":"'$latestOffset'"
    }'

>    3. In the application, using the ``FeaturedAppRight`` interface, exercise

Reviewer: Missing: the app's Daml code will have to depend on the `splice-api-featured-app-v1.dar`; and take an argument of type `ContractId FeaturedAppRight` on the choices whose execution should be featured, which allows the choice body to call the `FeaturedAppRight_CreateActivityMarker` choice to create the marker.

Author: Added.

Done.

   the ``FeaturedAppRight_CreateActivityMarker`` choice. Set the
   templateId to the fully qualified interface ID above.

   4. For testing examples, please review the example DamlScript test

>    `here <https://github.com/hyperledger-labs/splice/blob/main/daml/splice-dso-governance-test/daml/Splice/Scripts/TestFeaturedAppActivityMarkers.daml>`__.

Author: Done.  Learned about permalink today.

A non-featured app cannot accrue a ``FeaturedAppActivityMarker``.

Consider a single, simple transaction of a RWA which creates a single
``FeaturedAppActivityMarker`` activity record for one ``provider`` and the
``beneficiary`` is the ``provider``:

      1. A `FeaturedAppActivityMarker <https://docs.dev.sync.global/app_dev/api/splice-amulet/Splice-Amulet.html#type-splice-amulet-featuredappactivitymarker-16451>`__
         contract is created in the business transaction. The ``provider`` is
         set to the featured application provider's party. The ``beneficiary``
         must be set (unlike the
         `AppRewardCoupon <https://docs.dev.sync.global/app_dev/api/splice-amulet/Splice-Amulet.html#type-splice-amulet-apprewardcoupon-57229>`__)

>          to the party that would receive the CC. The ``provider`` field of the

Reviewer: ```suggestion
         to the party that should be eligible to mint the CC for that activity. The ``provider`` field of the
```

         `FeaturedAppActivityMarker <https://docs.dev.sync.global/app_dev/api/splice-amulet/Splice-Amulet.html#type-splice-amulet-featuredappactivitymarker-16451>`__
         is set by calling the interface choice
         `FeaturedAppRight_CreateActivityMarker <https://docs.dev.sync.global/app_dev/api/splice-api-featured-app-v1/Splice-Api-FeaturedAppRightV1.html#type-splice-api-featuredapprightv1-featuredapprightcreateactivitymarker-36646>`__.

      2. No ``ValidatorRewardCoupon`` is created.

There can be several ``FeaturedAppActivityMarkers`` per transaction tree

> which increases the total reward. It is also possible for a single

Reviewer: Missing: that this is only allowed for composed transactions like a settlement transaction where trading venue and all the registries of the transferred assets would get featured app rewards. 

This is not enforced in code. However, the tokenomics committee enforces fair usage of one's featured app right.

Author: Added. 

Done.

Canton transaction tree to include
`ValidatorRewardCoupon <https://docs.dev.sync.global/app_dev/api/splice-amulet/Splice-Amulet.html#type-splice-amulet-validatorrewardcoupon-76808>`__,
an
`AppRewardCoupon <https://docs.dev.sync.global/app_dev/api/splice-amulet/Splice-Amulet.html#type-splice-amulet-apprewardcoupon-57229>`__
and a
`FeaturedAppActivityMarker <https://docs.dev.sync.global/app_dev/api/splice-amulet/Splice-Amulet.html#type-splice-amulet-featuredappactivitymarker-16451>`__\ (s)
if there are sub-transcations that create each separately.

> It is possible to share the CC for the ``FeaturedAppActivityMarker``. The

Reviewer: It's not CC that is shared but the attribution of activity.
